chain/types/tipset_key.go

```diff
@@ -0,0 +1,98 @@
+package types
+
+import (
+	"bytes"
+	"encoding/binary"
+	"strings"
+
+	"github.com/ipfs/go-cid"
+)
+
+// A TipSetKey is an immutable collection of CIDs forming a unique key for a tipset.
+// The CIDs are assumed to be distinct and in canonical order. Two keys with the same
+// CIDs in a different order are not considered equal.
+// TipSetKey is a lightweight value type, and may be compared for equality with ==.
+type TipSetKey struct {
+	// The internal representation is a concatenation of the bytes of the CIDs, which are
+	// self-describing, wrapped as a string.
+	// These gymnastics make the a TipSetKey usable as a map key.
+	// The empty key has value "".
+	value string
+}
+
+// NewTipSetKey builds a new key from a slice of CIDs.
+// The CIDs are assumed to be ordered correctly.
+func NewTipSetKey(cids ...cid.Cid) TipSetKey {
+	encoded, err := encodeKey(cids)
+	if err != nil {
+		panic("failed to encode CIDs: " + err.Error())
+	}
+	return TipSetKey{string(encoded)}
+}
+
+// TipSetKeyFromBytes wraps an encoded key, validating correct decoding.
+func TipSetKeyFromBytes(encoded []byte) (TipSetKey, error) {
+	_, err := decodeKey(encoded)
+	if err != nil {
+		return TipSetKey{}, err
+	}
+	return TipSetKey{string(encoded)}, nil
+}
+
+// Cids returns a slice of the CIDs comprising this key.
+func (k TipSetKey) Cids() []cid.Cid {
+	cids, err := decodeKey([]byte(k.value))
+	if err != nil {
+		panic("invalid tipset key: " + err.Error())
+	}
+	return cids
+}
+
+// String() returns a human-readable representation of the key.
+func (k TipSetKey) String() string {
+	b := strings.Builder{}
+	b.WriteString("{")
+	cids := k.Cids()
+	for i, c := range cids {
+		b.WriteString(c.String())
+		if i < len(cids)-1 {
+			b.WriteString(",")
+		}
+	}
+	b.WriteString("}")
+	return b.String()
+}
+
+// Bytes() returns a binary representation of the key.
+func (k TipSetKey) Bytes() []byte {
+	return []byte(k.value)
+}
+
+func encodeKey(cids []cid.Cid) ([]byte, error) {
+	buffer := new(bytes.Buffer)
+	for _, c := range cids {
+		err := binary.Write(buffer, binary.LittleEndian, c.Bytes())
+		if err != nil {
+			return nil, err
+		}
+	}
+	return buffer.Bytes(), nil
+}
+
+func decodeKey(encoded []byte) ([]cid.Cid, error) {
+	// Estimate the number of CIDs to be extracted by dividing the encoded length by the shortest
+	// common CID length (V0 CIDs are 34 bytes). V1 CIDs are longer which means this might
+	// over-allocate, but avoid reallocation of the underlying array.
+	estimatedCount := len(encoded) / 34
```

Comment:
Block CIDs are 38 bytes, and given that this is what we are deserializing here, this should be:
```suggestion
	estimatedCount := len(encoded) / 38
```